### PR TITLE
New oxy with gorilla for websocket with integration tests

### LINKED ISSUE #1858: websocket inconsistency
### Do you want to request a *feature* or report a *bug*?
bug

### What did you do?

I first tried to get support on stackoverflow, but I think there's something else going on.

Post here for ref, but I'll restate below
https://stackoverflow.com/questions/44981861/websockets-in-traefik-opening-handshake-timed-out

I'm trying to get home-assistant in docker running behind traefik and I think I've had similar issues with other apps using websockets in the past (linuxserver/unifi controller and https://github.com/Division42LLC/division42llc-dotnet-webca/issues/2). The front page loads but when I try to login it just says connecting... and keeps spinning.

I open up the debugger in chrome and see this:

`WebSocket connection to 'ws://home-assistant.docker.int.xxx.com/api/websocket' failed: WebSocket opening handshake timed out
`

If I pass the port through directly to the host (i.e. http://docker.int.xxx.com:8123) it works.

### What did you expect to see?
websockets pass through successfully for app to work

### What did you see instead?
page loads but no connection to backend through websocket

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
v1.3.2, v1.3.3
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Docker in a Photon VM on esxi 5.5
`
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.8.1
 Git commit:   092cba3
 Built:        Fri May  5 02:08:33 2017
 OS/Arch:      linux/amd64
`

```sh
docker run -d --name="home-assistant" \
  --restart=always \
  -v $PWD/config:/config \
  -v /etc/localtime:/etc/localtime:ro \
  --network=traefik-net \
  -p 8123:8123 \
  --label traefik.port=8123 \
  --label traefik.docker.network=traefik-net \
  homeassistant/home-assistant
```

```toml
debug = true
logLevel = "debug"

defaultEntryPoints = ["http", "https", "ws", "wss"]
[entryPoints]
  [entryPoints.http]
  address = ":80"

  [entryPoints.https]
  address = ":443"
    [entryPoints.https.tls]

InsecureSkipVerify = true

#... acme and some manual backends ...

[docker]
endpoint = "unix:///var/run/docker.sock"
domain = "docker.int.xxx.com"
```

### If applicable, please paste the log output in debug mode (`--debug` switch)

```
time="2017-07-08T01:58:18Z" level=debug msg="Writing outgoing Websocket request to target connection: &{Method:GET URL:ws:/api/websocket Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Cache-Control:[no-cache] Connection:[Upgrade] User-Agent:[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36] Accept-Language:[en-US,en;q=0.8] Sec-Websocket-Version:[13] Cookie:[__zlcmid=eqgBjgiVtncb4A; _ga=GA1.2.930888288.1496028630] Sec-Websocket-Extensions:[permessage-deflate; client_max_window_bits] Accept-Encoding:[gzip, deflate] Origin:[http://home-assistant.docker.int.xxx.com] Upgrade:[websocket] Sec-Websocket-Key:[FsZu8u+pr9IQDAc3xZYiVQ==] Pragma:[no-cache]] Body:{} GetBody:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:home-assistant.docker.int.xxx.com Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr:192.168.2.1:49364 RequestURI:/api/websocket TLS:<nil> Cancel:<nil> Response:<nil> ctx:0xc4206ce4c0}"
```

### PR DESCRIPTION
**Pull request by Juliens**
- Use the new websocket implementation in oxy, based on gorilla/websocket
 - Add integration tests for verify websocket

fixes #1858 

In order to reproduce the issue:
```bash
docker run -d --name="home-assistant" \
    --restart=always \
    -v $PWD/config:/config \
    -v /etc/localtime:/etc/localtime:ro \
    -p 8123:8123 \
    --label traefik.port=8123 \
    --label traefik.frontend.rule="Host:127.0.0.1" \
    homeassistant/home-assistant
```
```
./traefik --docker=true
```

Then just try to view the homepage (http://127.0.0.1)

### CODE REVIEW

File: vendor/github.com/vulcand/oxy/forward/fwd.go [hunk 1]
```diff
@@ -335,6 +303,7 @@ func (f *websocketForwarder) copyRequest(req *http.Request, u *url.URL) (outReq
 
 	outReq.URL = utils.CopyURL(req.URL)
 	outReq.URL.Scheme = u.Scheme
+	outReq.URL.Path = outReq.RequestURI
```
Comment (wader) [inline vendor/github.com/vulcand/oxy/forward/fwd.go]: Hi, after upgrading to latest traefik (1.3.5) i noticed that query parameter sent in websock requests ends up URL-encoded on the backend server. Had a quick look and this line looks suspicious.

Websocket request:
```
curl -H "Connection: Upgrade" -H "Upgrade: websocket" -H "Sec-WebSocket-Version: 13" -H "Sec-WebSocket-Key: dummy" http://host/events?token=123
```
I see a requests to the backend that looks like this: (dumping with tcpdump -A)
```
GET /events%3Ftoken=123 HTTP/1.1     <-- notice the encoded "?"
Host: host
User-Agent: curl/7.54.0
Accept: */*
Connection: Upgrade
Sec-WebSocket-Key: 92tmVWFXOLFebYZc98UaAw==
Sec-WebSocket-Version: 13
Upgrade: websocket
X-Traefik-Reqid: 82
```

A normal GET request:
```
curl http://host/events?token=123
```
I see a request to the backend that looks like this:
```
GET /events?token=123 HTTP/1.1
Host: host
User-Agent: curl/7.54.0
Accept: */*
X-Forwarded-For: 1.2.3.4
X-Forwarded-Host: host
X-Forwarded-Proto: https
X-Forwarded-Server: 6ea428775acc
X-Traefik-Reqid: 87
Accept-Encoding: gzip
```

Now I also noticed that a lot of `X-Forwarded` headers are missing for the websocket request.

Comment (ldez) [inline vendor/github.com/vulcand/oxy/forward/fwd.go]: this line fixed the encoding pb.

Comment (Juliens) [inline vendor/github.com/vulcand/oxy/forward/fwd.go]: Thx for your report, the problem is already fix, just need to be merge and release
https://github.com/containous/oxy/pull/22

Comment (wader) [inline vendor/github.com/vulcand/oxy/forward/fwd.go]: @Juliens Ok! thanks for fix and info!

**Pull request merged**